### PR TITLE
Add X264 encoder

### PR DESCRIPTION
**Pull request by lherman-cs**
**Description**
* Add x264 codec (requires `libvpx-dev` to be installed)
* Support KeyFrameInterval
* Support Quality rate control

### CODE REVIEW

File: pkg/codec/x264/bridge.h [hunk 1]
```diff
@@ -0,0 +1,88 @@
...
+  e->param.b_annexb = 1;
+
+  if (x264_param_apply_profile(&e->param, "baseline") < 0)
+    goto fail;
+
+  if (x264_picture_alloc(&e->pic_in, param.i_csp, param.i_width, param.i_height) < 0)
+    goto fail;
+
+#undef fail
+#define fail fail2
```
Comment (at-wat) [inline pkg/codec/x264/bridge.h]: `fail` label is not used later. Just `undef` seems enough to avoid leaking.

File: pkg/codec/x264/bridge.h [hunk 2]
```diff
@@ -0,0 +1,88 @@
...
+    goto fail;
+
+  /* Configure non-default params */
+  e->param.i_csp = param.i_csp;
+  e->param.i_width = param.i_width;
+  e->param.i_height = param.i_height;
+  e->param.i_fps_num = param.i_fps_num;
+  e->param.i_fps_den = 1;
+  // Intra refres:
+  e->param.i_keyint_max = param.i_fps_num;
```
Comment (at-wat) [inline pkg/codec/x264/bridge.h]: https://github.com/pion/mediadevices/blob/6fd00ab14568e16971363b1622ba58c9feb9b15b/pkg/prop/prop.go#L128
`Codec.KeyFrameInterval` can be used.

File: pkg/codec/x264/bridge.h (hunk 1 above)
Comment (lherman-cs) [reply pkg/codec/x264/bridge.h]: I've removed it, instead uses fail2 directly instead.

File: pkg/codec/x264/bridge.h (hunk 2 above)
Comment (lherman-cs) [reply pkg/codec/x264/bridge.h]: Oops, makes sense. Updated it.

**Comment by lherman-cs** [comment]
@at-wat I've updated the PR. PTAL.

**Comment by lherman-cs** [comment]
It seems like we need a better way of controlling the rate control. For example, `Bitrate` and `Quality` seem to be mutually exclusive, `Quality` impacts `Bitrate` and also the other way around.

**Comment by at-wat** [comment]
> For example, Bitrate and Quality seem to be mutually exclusive, Quality impacts Bitrate and also the other way around.

In codecs, quality is separately configurable. It's usually a complexity of inter frame prediction. By using more CPU power to it, same size of the compressed data can represents larger amount of information.

**Comment by lherman-cs** [comment]
Wait, so am I interpreting the RF parameter wrong? I'm using the Quality to control RF.

**Comment by at-wat** [comment]
I'm not very sure, but in 
https://github.com/mirror/x264/blob/1771b556ee45207f8711744ccbd5d42a3949b14c/x264.h#L441-L442
`f_rf_constant` seems like a VBR control parameter and `i_bitrate` also exists.

**Comment by lherman-cs** [comment]
So, apparently, if we choose to use CRF mode, `x264` will reset `i_bitrate` to 0 (ignore), https://github.com/mirror/x264/blob/1771b556ee45207f8711744ccbd5d42a3949b14c/encoder/encoder.c#L874.

File: pkg/codec/x264/bridge.h [hunk 3]
```diff
@@ -0,0 +1,94 @@
...
+  /* Configure non-default params */
+  e->param.i_csp = param.i_csp;
+  e->param.i_width = param.i_width;
+  e->param.i_height = param.i_height;
+  e->param.i_fps_num = param.i_fps_num;
+  e->param.i_fps_den = 1;
+  // Intra refres:
+  e->param.i_keyint_max = param.i_keyint_max;
+  // Rate control:
+  e->param.rc.i_rc_method = X264_RC_CRF;
```
Comment (at-wat) [inline pkg/codec/x264/bridge.h]: I think VBR mode is better especially for WebRTC.
Bitrate of CRF depends on the input video and takes high bitrate if entropy of the input is large.

**Comment by at-wat** [comment]
Encoding complexity can be configured like https://github.com/mirror/x264/blob/1771b556ee45207f8711744ccbd5d42a3949b14c/common/base.c#L434-L540.
Slower encoding let output quality higher for same bitrate.

**Comment by lherman-cs** [comment]
According to this article https://slhck.info/video/2017/03/01/rate-control.html, both ABR and CRF seem to be fine rate controls. But, what I'm missing here is VBV-constained bitrate to make sure that we're encoding under the bandwidth limit.

**Comment by lherman-cs** [comment]
Looking at this block of code https://github.com/mirror/x264/blob/master/encoder/encoder.c#L1838-L1848, it seems that we only need to set `i_vbv_max_bitrate` and `i_vbv_buffer_size` to be greater than zero

**Comment by lherman-cs** [comment]
> According to this article https://slhck.info/video/2017/03/01/rate-control.html, both ABR and CRF seem to be fine rate controls. But, what I'm missing here is VBV-constained bitrate to make sure that we're encoding under the bandwidth limit.

After a second thought, ABR seems to be more suitable for WebRTC like you said. As the goal in WebRTC is usually to meet a bitrate target, we can use ABR to control it based on the average bitrate to meet this bitrate target from the user.

Then, finally, we can get some help from VBV constraint to meet the target bitrate by using a buffer.

**Comment by lherman-cs** [comment]
@at-wat I just updated the PR to use ABR+VBV. PTAL.

File: pkg/codec/x264/x264.go [hunk 4]
```diff
@@ -0,0 +1,159 @@
...
+		i_width:      C.int(p.Width),
+		i_height:     C.int(p.Height),
+		i_keyint_max: C.int(p.KeyFrameInterval),
+	}
+	param.rc.i_bitrate = C.int(p.BitRate)
+	param.rc.i_vbv_max_bitrate = param.rc.i_bitrate
+	param.rc.i_vbv_buffer_size = param.rc.i_vbv_max_bitrate * 2
+
+	var rc C.int
+	cPreset := C.CString(fmt.Sprint(preset))
```
Comment (at-wat) [inline pkg/codec/x264/x264.go]: maybe nice to have a comment that `cPreset` is freed in `enc_new`.

Comment (lherman-cs) [reply pkg/codec/x264/x264.go]: ah, yep

**Pull request merged**